### ISSUE TITLE
Issues when compiling

### ISSUE THREAD

**Issue by 0xfadead** [facebook/igl#178]
I'm trying to compile IGL with default settings.
I ran the following to build it:
```
$ ./deploy_dependencies.py
$ ./deplay_content.py
$ cmake -Bbuild -S. -GUnix\ Makefiles -DCMAKE_EXPORT_COMPILE_COMMANDS=ON
$ make -C build -j10
```

I am on Linux with the following `uname -a`: `Linux [HOSTNAME] 6.10.10-arch1-1 #1 SMP PREEMPT_DYNAMIC Thu, 12 Sep 2024 17:21:02 +0000 x86_64 GNU/Linux`
I have both GCC and Clang installed, it seems to have been using GCC though, because when running with clang it gave me a bunch of extra warnings about pragmas.
GCC version:
```
$ gcc --version
gcc (GCC) 14.2.1 20240910
Copyright (C) 2024 Free Software Foundation, Inc.
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
```
Clang version:
```
$ clang --version
clang version 18.1.8
Target: x86_64-pc-linux-gnu
Thread model: posix
InstalledDir: /usr/bin
```

`cmake -Bbuild -S. -DCMAKE_EXPORT_COMPILE_COMMANDS -GUnix\ Makefiles` Output:
```
. . .
CMake Warning (dev) at third-party/deps/src/glslang/CMakeLists.txt:236 (find_package):
  Policy CMP0148 is not set: The FindPythonInterp and FindPythonLibs modules
  are removed.  Run "cmake --help-policy CMP0148" for policy details.  Use
  the cmake_policy command to set the policy and suppress this warning.

Call Stack (most recent call first):
  third-party/deps/src/glslang/CMakeLists.txt:300 (find_host_package)
This warning is for project developers.  Use -Wno-dev to suppress it.
. . .
CMake Deprecation Warning at third-party/deps/src/bc7enc/CMakeLists.txt:3 (cmake_minimum_required):
  Compatibility with CMake < 3.5 will be removed from a future version of
  CMake.

  Update the VERSION argument <min> value or use a ...<max> suffix to tell
  CMake that the project does not need compatibility with older versions.


Initial BUILD_X64=ON
Initial CMAKE_BUILD_TYPE=
bc7enc build type: Release
Building 64-bit
CMake Deprecation Warning at third-party/deps/src/tinyobjloader/CMakeLists.txt:4 (cmake_minimum_required):
  Compatibility with CMake < 3.5 will be removed from a future version of
  CMake.

  Update the VERSION argument <min> value or use a ...<max> suffix to tell
  CMake that the project does not need compatibility with older versions.
. . .
-- Version: 10.2.1
-- Build type:
-- Found Bash: /usr/bin/bash
--   AVX2 backend     - ON
--   SSE4.1 backend   - OFF
--   SSE2 backend     - OFF
--   NEON backend     - OFF
--   NONE backend     - OFF
--   NATIVE backend   - OFF
--   Invariance       - ON
--   Shared libs      - OFF
--   Decompressor     - OFF
--   Diagnostics      - OFF
--   ASAN             - OFF
--   Unit tests       - OFF
-- Configuring done (1.1s)
-- Generating done (0.2s)
. . .
```

`make -C build -j10` Output:
```
. . .
[ 66%] Building CXX object IGLU/CMakeFiles/IGLUtexture_loader.dir/texture_loader/ktx1/TextureLoaderFactory.cpp.o
/home/doomdot/Projects/source_build/igl/igl/IGLU/uniform/Collection.cpp: In member function ‘void iglu::uniform::Collection::clear(const igl::NameHandle&)’:
/home/doomdot/Projects/source_build/igl/igl/IGLU/uniform/Collection.cpp:32:40: error: cannot convert ‘std::vector<igl::NameHandle>::iterator’ to ‘const char*’
   32 |   names_.erase(std::remove(names_.begin(), names_.end(), name), names_.end());
      |                            ~~~~~~~~~~~~^~
      |                                        |
      |                                        std::vector<igl::NameHandle>::iterator
In file included from /usr/include/c++/14.2.1/cstdio:42,
                 from /usr/include/c++/14.2.1/ext/string_conversions.h:45,
                 from /usr/include/c++/14.2.1/bits/basic_string.h:4154,
                 from /usr/include/c++/14.2.1/string:54,
                 from /home/doomdot/Projects/source_build/igl/igl/src/igl/Common.h:19,
                 from /home/doomdot/Projects/source_build/igl/igl/src/igl/Buffer.h:11,
                 from /home/doomdot/Projects/source_build/igl/igl/./IGLU/uniform/Trait.h:12,
                 from /home/doomdot/Projects/source_build/igl/igl/./IGLU/uniform/Descriptor.h:10,
                 from /home/doomdot/Projects/source_build/igl/igl/./IGLU/uniform/Collection.h:10,
                 from /home/doomdot/Projects/source_build/igl/igl/IGLU/uniform/Collection.cpp:8:
/usr/include/stdio.h:158:32: note:   initializing argument 1 of ‘int remove(const char*)’
  158 | extern int remove (const char *__filename) __THROW;
      |                    ~~~~~~~~~~~~^~~~~~~~~~
make[2]: *** [IGLU/CMakeFiles/IGLUuniform.dir/build.make:76: IGLU/CMakeFiles/IGLUuniform.dir/uniform/Collection.cpp.o] Error 1
make[2]: *** Waiting for unfinished jobs....
. . .[OTHER UNFINISHED JOBS]. . .
```

If I fix this compilation error by adding `#include <algorithm>` to the offending file, I get more compilation errors:
`make -C build -j10` Output:
```
. . .
[ 81%] Building CXX object shell/CMakeFiles/BasicFramebufferSession_opengl.dir/renderSessions/BasicFramebufferSession.cpp.o
In file included from /home/doomdot/Projects/source_build/igl/igl/shell/apps/SessionApp.cpp:14:
/home/doomdot/Projects/source_build/igl/igl/./shell/renderSessions/GPUStressSession.h:102:21: error: field ‘forceReset_’ has incomplete type ‘std::atomic<bool>’
  102 |   std::atomic<bool> forceReset_{false};
      |                     ^~~~~~~~~~~
In file included from /usr/include/c++/14.2.1/bits/shared_ptr_atomic.h:33,
                 from /usr/include/c++/14.2.1/memory:81,
                 from /home/doomdot/Projects/source_build/igl/igl/./shell/shared/renderSession/RenderSession.h:10,
                 from /home/doomdot/Projects/source_build/igl/igl/./shell/renderSessions/GPUStressSession.h:12:
/usr/include/c++/14.2.1/bits/atomic_base.h:172:12: note: declaration of ‘struct std::atomic<bool>’
  172 |     struct atomic;
      |            ^~~~~~
make[2]: *** [shell/CMakeFiles/GPUStressSession_opengl.dir/build.make:76: shell/CMakeFiles/GPUStressSession_opengl.dir/apps/SessionApp.cpp.o] Error 1
make[1]: *** [CMakeFiles/Makefile2:2005: shell/CMakeFiles/GPUStressSession_opengl.dir/all] Error 2
make[1]: *** Waiting for unfinished jobs....
[ 82%] Building CXX object shell/CMakeFiles/DrawInstancedSession_opengl.dir/renderSessions/DrawInstancedSession.cpp.o
[ 83%] Building CXX object shell/CMakeFiles/MRTSession_opengl.dir/renderSessions/MRTSession.cpp.o
/home/doomdot/Projects/source_build/igl/igl/shell/renderSessions/ColorSession.cpp: In member function ‘virtual void igl::shell::ColorSession::initialize()’:
/home/doomdot/Projects/source_build/igl/igl/shell/renderSessions/ColorSession.cpp:246:3: error: ‘memcpy’ was not declared in this scope
  246 |   memcpy(&fragmentParameters_.mvp, &mvp, sizeof(mvp));
      |   ^~~~~~
/home/doomdot/Projects/source_build/igl/igl/shell/renderSessions/ColorSession.cpp:17:1: note: ‘memcpy’ is defined in header ‘<cstring>’; this is probably fixable by adding ‘#include <cstring>’
   16 | #include <shell/shared/renderSession/ShellParams.h>
  +++ |+#include <cstring>
   17 |
/home/doomdot/Projects/source_build/igl/igl/shell/renderSessions/ColorSession.cpp:247:48: error: cannot convert ‘<brace-enclosed initializer list>’ to ‘iglu::simdtypes::float3’ {aka ‘__vector(4) float’} in assignment
  247 |   fragmentParameters_.color = {1.0f, 1.0f, 1.0f};
      |                                                ^
/home/doomdot/Projects/source_build/igl/igl/shell/renderSessions/ColorSession.cpp: In member function ‘virtual void igl::shell::ColorSession::update(igl::SurfaceTextures)’:
/home/doomdot/Projects/source_build/igl/igl/shell/renderSessions/ColorSession.cpp:329:3: error: ‘memcpy’ was not declared in this scope
  329 |   memcpy(&fragmentParameters_.mvp, &mvp, sizeof(mvp));
      |   ^~~~~~
/home/doomdot/Projects/source_build/igl/igl/shell/renderSessions/ColorSession.cpp:329:3: note: ‘memcpy’ is defined in header ‘<cstring>’; this is probably fixable by adding ‘#include <cstring>’
. . .
```

I can fix the first error by adding `#include <atomic>` in `GPUStressSession.h` and I can fix the second issue by adding `#include <cstring>` to `ColorSession.cpp`

Now I get another new error (yay!):
`make -C build -j10` Output:
```
. . .
[ 88%] Building CXX object shell/CMakeFiles/Textured3DCubeSession_opengl.dir/apps/SessionApp.cpp.o
/home/doomdot/Projects/source_build/igl/igl/shell/renderSessions/ColorSession.cpp: In member function ‘virtual void igl::shell::ColorSession::initialize()’:
/home/doomdot/Projects/source_build/igl/igl/shell/renderSessions/ColorSession.cpp:249:48: error: cannot convert ‘<brace-enclosed initializer list>’ to ‘iglu::simdtypes::float3’ {aka ‘__vector(4) float’} in assignment
  249 |   fragmentParameters_.color = {1.0f, 1.0f, 1.0f};
      |                                                ^
make[2]: *** [shell/CMakeFiles/ColorSession_vulkan.dir/build.make:90: shell/CMakeFiles/ColorSession_vulkan.dir/renderSessions/ColorSession.cpp.o] Error 1
make[1]: *** [CMakeFiles/Makefile2:1745: shell/CMakeFiles/ColorSession_vulkan.dir/all] Error 2
make[1]: *** Waiting for unfinished jobs....
[ 88%] Building CXX object shell/CMakeFiles/Textured3DCubeSession_opengl.dir/renderSessions/Textured3DCubeSession.cpp.o
/home/doomdot/Projects/source_build/igl/igl/shell/renderSessions/ColorSession.cpp: In member function ‘virtual void igl::shell::ColorSession::initialize()’:
/home/doomdot/Projects/source_build/igl/igl/shell/renderSessions/ColorSession.cpp:249:48: error: cannot convert ‘<brace-enclosed initializer list>’ to ‘iglu::simdtypes::float3’ {aka ‘__vector(4) float’} in assignment
  249 |   fragmentParameters_.color = {1.0f, 1.0f, 1.0f};
      |                                                ^
[ 88%] Linking CXX executable ImguiSession_vulkan
make[2]: *** [shell/CMakeFiles/ColorSession_opengl.dir/build.make:90: shell/CMakeFiles/ColorSession_opengl.dir/renderSessions/ColorSession.cpp.o] Error 1
make[1]: *** [CMakeFiles/Makefile2:1797: shell/CMakeFiles/ColorSession_opengl.dir/all] Error 2
. . .
```

I have no idea how to fix that issue and I don't think I should have to fix that many compilation errors just to compile with the default settings.
What am I doing wrong?

Thanks in advance!

**Comment by corporateshark** [comment]
@0xfadead This [commit](https://github.com/facebook/igl/commit/35e0cce49427ef02e0246825e7d35817a07afeaa) fixes a bunch of Linux compilation issues.

I was not able to reproduce the issue with `fragmentParameters_.color` in my environment. Do you have any possibility to use previous versions of Clang/GCC?

**Comment by corporateshark** [comment]
It's been over a week since we last heard from the OP. I am closing this issue as "cannot reproduce."

**Comment by corporateshark** [comment]
More Linux issues fixed: 59de6782c78a763c0483e92ed0a9dbe8553ecc36
